src/internal/futex/futex.go

```diff
@@ -0,0 +1,74 @@
+package futex
+
+// Cross platform futex implementation.
+// Futexes are supported on all major operating systems and on WebAssembly.
+//
+// For more information, see: https://outerproduct.net/futex-dictionary.html
+
+import (
+	"sync/atomic"
+	"unsafe"
+)
+
+// A futex is a way for userspace to wait with the pointer as the key, and for
+// another thread to wake one or all waiting threads keyed on the same pointer.
+//
+// A futex does not change the underlying value, it only reads it before going
+// to sleep (atomically) to prevent lost wake-ups.
+type Futex struct {
+	atomic.Uint32
+}
+
+// Atomically check for cmp to still be equal to the futex value and if so, go
+// to sleep. Return true if we were definitely awoken by a call to Wake or
+// WakeAll, and false if we can't be sure of that.
+func (f *Futex) Wait(cmp uint32) bool {
+	tinygo_futex_wait((*uint32)(unsafe.Pointer(&f.Uint32)), cmp)
```

Comment:
It's probably ok to assume the layout of `atomic.Uint32` here and below, but it seems to me you can avoid it by

```
type Futex struct {
    v uint32
}
```

and adding methods that forwards to atomic.(Load|And|Store...)Uint32 functions.

Reply:
I'd prefer to keep using `atomic.Uint32`. I don't expect `atomic.Uint32` to ever change to something else, and if it does we can always update this code.